docs/monorepo/engine_binary_hashing.md

```diff
@@ -0,0 +1,96 @@
+11a5a03d15ae21bde366e41291a7899eec44e5ae  -
+
+git ls-tree -r HEAD  engine/src/flutter/vulkan/vulkan_window.h
+100644 blob 11a5a03d15ae21bde366e41291a7899eec44e5ae	engine/src/flutter/vulkan/vulkan_window.h
+```
+
+## Scoping the Hash to the Engine
+
+To accurately track engine binaries, we only want to include files that directly contribute to the engine build. This includes the `engine/` directory and the root `DEPS` file, which tracks third-party dependencies managed by `gclient sync`. Using `git ls-tree -r HEAD engine DEPS` effectively captures all necessary files while excluding irrelevant content from the `third_party` directory.
+
+```shell
+100644 blob 5143313ce5826665309e8a086a281ad3ab1a9ce7    DEPS
+100644 blob 205edfe43306c4dbf9a4a6f15e83cf5d49b9fc7d    engine/src/flutter/.ci.yaml
+100644 blob 3c73f32a334086d9a0f4fd468dcdf9505d74e9c5    engine/src/flutter/.clang-format
+100644 blob b74be267bc42f08ebf9afe8eec5cbbfe75c5a1c9    engine/src/flutter/.clang-tidy
+100644 blob dd395bfd2104526d4f865313eab578f15ee5775b    engine/src/flutter/.engine-release.version
+100644 blob e69de29bb2d1d6434b8b29ae775ad8c2e48c5391    engine/src/flutter/.git-blame-ignore-revs
+100644 blob 915d1ed51d121f1986c9dfe71cf1745c1a11286d    engine/src/flutter/.gitattributes
+100644 blob c1c1d3d05f37b0e09155b32aceb6d2ec62ee464b    engine/src/flutter/.github/PULL_REQUEST_TEMPLATE.md
+100644 blob 9688ddae25af122d7c17d9c27d887b84888f3619    engine/src/flutter/.github/dependabot.yml
+100644 blob ed7171a9638274d8f411b6bededec61feab15a7b    engine/src/flutter/.github/labeler.yml
+100644 blob be245c915e7eb5377317cc6eb038442628071790    engine/src/flutter/.github/release.yml
+# ... all files
+```
+
+To generate a consistent hash across different platforms (including Windows CI environments), we can use `git hash-object`:
+
+```bash
+git ls-tree -r HEAD engine DEPS | git hash-object --stdin
+3b9abe00dec28902a589c982b5b460b0f9f38e93
+```
+
+## Supporting A/B Testing
+
+When developing a pull request (PR), your branch might contain multiple commits. To enable A/B testing against the engine version at the time of branching, we can modify the hash calculation to use the merge-base. This ensures that the generated hash reflects the engine state at the branch point, facilitating accurate comparisons.
+
+```bash
+git ls-tree -R $(git merge-base HEAD master) engine DEPS | git hash-object --stdin
+```
+
+## Considerations and Future Refinements
+
+Currently, `git ls-tree` incorporates file names and permissions into the hash calculation. Consequently, renaming or changing permissions of a file will necessitate rebuilding the engine. While acceptable initially, this behavior might require adjustments in the future.
+
+If we want to focus solely on file content and size, we could use `git ls-tree -r --object-only engine DEPS | sort | git hash-object --stdin`. However, this relies on consistent sorting across operating systems, which might introduce complexities.
```

Comment:
Sorry, just skimming so maybe you address this, but I'm noticing here that the output of `ls-tree` is sorted before handing to `git hash-object`, but *isn't* sorted below in the command down on line 93.

Reply:
This section is talking about tuning the output of ls-tree to only output the Sha1 sum and then sorting it. That should make the hash resilient to files moving and being renamed.

However, this use case is fairly specialized and potentially rare. It's probably not worth the complication right now.

Reply:
Line 93 has the internal sort for git trees.

Reply:
Got it. Maybe `Considerations and Future Refinements` should come after `Recommended Formula and Implementation`.

Reply:
Sure thing